### ISSUE TITLE
templates evaluate at global scope, breaks expression`return;`

### ISSUE THREAD

**Issue by pieterv** [facebook/jscodeshift#42]
Return statements can only be added inside a function but expressions are evaluated by babel at the global scope so babel throws on 
```
expression`return;`
```
Example: http://felix-kling.de/esprima_ast_explorer/#/1lLYuPcGmT

This also happens for `break;` statements as well.

@fkling had an idea of always evaluating templates within a function but that could break things like `import` that can only be added at the global scope. Not sure what the way forward is here but i find the templating extremely valuable so it would be great to find a solution!

**Comment by cpojer** [comment]
you can also use ast-types to build up your return statement. 

```js
j.returnStatement();
```

See https://github.com/benjamn/ast-types/blob/master/def/core.js#L106

**Comment by fkling** [comment]
FWIW

```
expression`return;`
```

would never work anyway, since a `return` statement is not an expression. but 

```
statement`return;`
```

has the same problem.

**Comment by fkling** [comment]
But yeah, in same cases it might just be easier to mix and match. E.g. instead of

```js
statement`return foo(${someNode});`
```

you could write

```js
j.returnStatement(expression`foo(${someNode})`)
```

**Comment by pieterv** [comment]
Sorry yes, statement not expression.

Yeah i have been mixing and matching but this:
```
statements`
var aasd = () => 'some complex stuff';
${j.returnStatement(expression`aasd(${someNode})`)}
`;
```

Is considerably less readable than this:
```
statements`
var aasd = () => 'some complex stuff';
return aasd(${someNode});
`;
```

I understand this is a hard problem to solve but i do think having a solution in the future will be helpful since it reduces the complexity of the code which is valuable as the transforms get more complex.

**Comment by cpojer** [comment]
I think we could potentially do this: try to parse it in the global scope. If that throws, try to parse it wrapped in a function and pull it out.

It seems a lot like

```js
var div = document.createElement('div');
div.innerHTML = text;
var myElement = div.childNodes[0];
```

:D

**Comment by pieterv** [comment]
Yeah that seems like it might work, i was wondering if there was a away to inform template what context it should run in by passing it a `path` or something.

**Comment by fkling** [comment]
yeah, the above solution would't work for things like `break`.

**Comment by fkling** [comment]
Maybe we can have a type passed as context. E.g.

```
statement.in(j.IfStatement)`break;`
```

it's a bit magical because the implication would be that the code is evaluated in the **body** of an `if` statement and it wouldn't work for all types, but it may the most concise way. 

----

Potentially supported types:

- Function
- IfStatement
- ForStatement, ForInStatement, WhileStatement, DoWhileStatement, ForOfStatement
- SwitchStatement
- SwitchCase

**Comment by pieterv** [comment]
I was just writing out the same suggestion! :P I think it will work, like you said we will need to manually code the cases but it shouldn't be to much overhead.
